motif/matchComp.go

```diff
@@ -0,0 +1,288 @@
+		scanAltSequenceComp(s.Records, revKmerHash, revCompMotif, s.ChromName, out, s.ResidualWindowSize, consensusScore, bed.Negative, s.RefStart, s.EnforceStrandMatch, s.OutputAsProportion, altEndsConsidered, s.ResidualFilter)
+	}
+
+	err = out.Close()
+	exception.PanicOnErr(err)
+}
+
+func scanRefSequenceComp(records []fasta.Fasta, kmerHash map[uint64]float64, pm PositionMatrix, chromName string, out *fileio.EasyWriter, residualWindowSize int, consensusScore float64, strand bed.Strand, refStart int, enforceStrandMatch bool, outputAsProportion bool, altEndsConsidered map[int]bool, residualFilter float64) {
+	var needNewKey bool = true
+	var currRefKey uint64
+	var couldGetNewKey, inKmerHash, couldScoreSequence bool
+	var currRefScore, currAltScore, currResidual, minResidual, minResidualAltScore float64
+	var refPos, lastRefPos, lastAlnPos, currAltStart, currAltEnd int = refStart, refStart, 0, 0, 0
+	var bitMask uint64 = uint64(math.Pow(2, float64(2*len(pm.Mat[0]))) - 1) // bitmask formula: B_n = 2^{2n} - 1
+	var currBed bed.Bed
+	revCompPm := ReverseComplement(pm)
+
+	// scan through reference sequence first
+	for alnPos := 0; alnPos < len(records[0].Seq); alnPos++ {
+		refPos = fasta.AlnPosToRefPosCounter(records[0], alnPos, lastRefPos, lastAlnPos)
+		//first we need a key for the alnPos. We can either make a new key or update our current key
+		if needNewKey {
+			currRefKey, alnPos, couldGetNewKey = getNewKey(records[0], alnPos, len(pm.Mat[0]))
+			refPos = fasta.AlnPosToRefPosCounter(records[0], alnPos, lastRefPos, lastAlnPos)
+			lastRefPos, lastAlnPos = refPos, alnPos
+			if !couldGetNewKey {
+				break //this means we've run out of windows on the reference sequence.
+			}
+			needNewKey = false
+		} else {
+			switch records[0].Seq[alnPos] {
+			case dna.N:
+				needNewKey = true
+				continue
+			case dna.Gap:
+				continue
+			case dna.A:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 0
+				currRefKey = currRefKey & bitMask
+			case dna.C:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 1
+				currRefKey = currRefKey & bitMask
+			case dna.G:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 2
+				currRefKey = currRefKey & bitMask
+			case dna.T:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 3
+				currRefKey = currRefKey & bitMask
+			default:
+				log.Fatalf("Unrecognized base.")
+			}
+		}
+
+		// now we check if the current key is a significant motif hit
+		if currRefScore, inKmerHash = kmerHash[currRefKey]; inKmerHash {
+			//fmt.Printf("CurrRefKey: %v. CurrRefScore: %v.\n", currRefKey, currRefScore)
+			minResidual = math.Inf(1)
+			minResidualAltScore = 0
+			for currAltStart = numbers.Max(alnPos-len(pm.Mat[0])-residualWindowSize+1, 0); currAltStart <= alnPos+residualWindowSize-len(pm.Mat[0])+1; currAltStart++ {
```

Comment:
would the for loop's end condition also need a numbers.Min between the "lnPos+residualWindowSize-len(pm.Mat[0])+1" and the end/length of the multiFa sequence?

Reply:
Great catch, I'll change at 243 as well